test/unit/alignment/matrix/detail/aligned_sequence_builder_test.cpp

```diff
@@ -0,0 +1,300 @@
+
+    two_dimensional_matrix<trace_directions> matrix{number_rows{3}, number_cols{4}, std::vector
+    {
+        N,           LO, L,          L,
+        UO, D | LO | UO, L, D | L | UO,
+        U,       LO | U, D,          L
+    }};
+
+    using fst_seq_t = std::tuple_element_t<0, test_type>;
+    using sec_seq_t = std::tuple_element_t<1, test_type>;
+
+    using fst_seq_value_t = value_type_t<fst_seq_t>;
+    using sec_seq_value_t = value_type_t<sec_seq_t>;
+
+    using type_param = aligned_sequence_builder<fst_seq_t, sec_seq_t>;
+
+    void SetUp()
+    {
+        fst.push_back(assign_char_to('A', fst_seq_value_t{}));
+        fst.push_back(assign_char_to('C', fst_seq_value_t{}));
+        fst.push_back(assign_char_to('G', fst_seq_value_t{}));
+        sec.push_back(assign_char_to('A', sec_seq_value_t{}));
+        sec.push_back(assign_char_to('G', sec_seq_value_t{}));
+
+        builder = type_param{fst, sec};
+    }
+
+    auto path(matrix_offset const & offset)
+    {
+        using iterator_t = decltype(trace_iterator{matrix.begin() + offset});
+        return std::ranges::subrange<iterator_t, std::ranges::default_sentinel_t>
+        {
+            trace_iterator{matrix.begin() + offset},
+            std::ranges::default_sentinel
+        };
+    }
+
+    std::remove_reference_t<fst_seq_t> fst;
+    std::remove_reference_t<sec_seq_t> sec;
+    type_param builder;
+};
+
+using test_types = ::testing::Types<std::pair<dna4_vector &, dna4_vector &>,
+                                    std::pair<dna4_vector &, dna15_vector &>,
+                                    std::pair<dna4_vector &, std::list<dna4> &>>;
+
+TYPED_TEST_CASE(aligned_sequence_builder_fixture, test_types);
+
+TYPED_TEST(aligned_sequence_builder_fixture, construction)
+{
+    EXPECT_TRUE(std::is_default_constructible_v<typename TestFixture::type_param>);
+    EXPECT_TRUE(std::is_copy_constructible_v<typename TestFixture::type_param>);
+    EXPECT_TRUE(std::is_move_constructible_v<typename TestFixture::type_param>);
+    EXPECT_TRUE(std::is_copy_assignable_v<typename TestFixture::type_param>);
+    EXPECT_TRUE(std::is_move_assignable_v<typename TestFixture::type_param>);
+    EXPECT_TRUE(std::is_destructible_v<typename TestFixture::type_param>);
+    EXPECT_TRUE((std::is_constructible_v<typename TestFixture::type_param,
+                                         typename TestFixture::fst_seq_t,
+                                         typename TestFixture::sec_seq_t>));
+}
+
+TYPED_TEST(aligned_sequence_builder_fixture, build_from_2_3)
+{
+    auto p = this->path(matrix_offset{row_index_type{2}, column_index_type{3}});
+    auto [begin, end, alignment] = this->builder.build_from(p);
+
+    EXPECT_EQ(begin.row, 0u);
+    EXPECT_EQ(begin.col, 0u);
+    EXPECT_EQ(end.row, 2u);
+    EXPECT_EQ(end.col, 3u);
+    EXPECT_EQ(alignment.first  | view::to_char | std::ranges::to<std::string>, std::string{"--ACG"});
```

Comment:
Mhm, I thought you'd get `L,D,D` from the trace back, and the result would be
```
ACG
AG-
```

Reply:
Nop, it models affine gaps. That means you need to follow a gap until you find the corresponding gap_open signal. With that respect linear gaps are a specialisation of affine gaps in which every gap is a gap open. Thus we can use the same trace algorithm for affine and linear gaps.